grant-project/zcash-grant-system
# Cancel proposal

Closes #244.

### What This Does

Builds off of the failed proposal flow to implement manual cancels, and fixes up a bunch of things related to canceled / failed proposals:

* Renames the `REFUNDING` stage to `FAILED`, and also adds a `CANCELED` stage. These are largely treated the same, but provide explanations more suited to why refunds will be happening. Includes a migration to change all proposals in `REFUNDING` to `FAILED`.
* Adds a button to the admin that allows an admin to cancel a proposal. This emails the team and all contributors in a way similar to failed proposals.
* Removes the "Edit proposal" action that team members saw, but was always greyed out on proposals.
* Reworks to "Cancel proposal" action to pop open a modal that lets the user know they'll need to contact support.
* Adds `stage` tags to `LIVE` proposals in the admin.
* Prevents `CANCELED` and `FAILED` proposals from showing up on the proposals list view. They can still be visited directly through links, they just won't show up in the main proposal browser.

### Steps to Test

* Create a new proposal with an immediately paid milestone, and get it to LIVE
* Fully fund the proposal as any user before the deadline is up
* Mark milestone 1 as paid with a txid
* Cancel the proposal to initiate the refunds, confirm both the proposal team and contributors get emails
* Confirm the admin shows a non-100% amount that is to be refunded
* Confirm the non staking contribution shows up as needing a refund, and shows the right amount to refund.

## Screenshots

### Cancel button

<img width="708" alt="screen shot 2019-02-22 at 2 06 26 pm" src="https://user-images.githubusercontent.com/649992/53264994-0f915100-36ab-11e9-8b2e-e7364f07c1be.png">


### Proposal canceled alert

<img width="913" alt="screen shot 2019-02-22 at 1 26 12 pm" src="https://user-images.githubusercontent.com/649992/53264973-01dbcb80-36ab-11e9-9f80-06db230d3a80.png">


### Partial contribution refund

<img width="918" alt="screen shot 2019-02-22 at 1 26 32 pm" src="https://user-images.githubusercontent.com/649992/53264962-fa1c2700-36aa-11e9-82fc-80a806527d9c.png">


### Canceled proposal campaign block

<img width="323" alt="screen shot 2019-02-22 at 1 46 11 pm" src="https://user-images.githubusercontent.com/649992/53264945-f25c8280-36aa-11e9-953c-ed951502fd3f.png">


### Cancel modal

<img width="575" alt="screen shot 2019-02-22 at 1 59 12 pm" src="https://user-images.githubusercontent.com/649992/53264936-ecff3800-36aa-11e9-8355-c7e135d42dfa.png">


### Stage tags in admin

<img width="443" alt="screen shot 2019-02-22 at 2 01 29 pm" src="https://user-images.githubusercontent.com/649992/53264924-e4a6fd00-36aa-11e9-8bcf-6ad71d952e25.png">
